File: LeetCode/LeetCode101_150/125_Valid_Palindrome.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    public static partial class LeetCode101_150
    {
        public static bool IsPalindrome(string s)
        {
            if (string.IsNullOrWhiteSpace(s))
            {
                return true;
            }
            int left = 0, right = s.Length - 1;
            while (left < right)
            {
                while (!char.IsLetterOrDigit(s[left]))
                {
                    left++;
                    if (left >= right)
                    {
                        return true;
                    }
                }
                while (!char.IsLetterOrDigit(s[right]))
                {
                    right--;
                    if (right <= left)
                    {
                        return true;
                    }
                }
                if (char.ToLowerInvariant(s[left]) != char.ToLowerInvariant(s[right]))
                {
                    return false;
                }
                left++;
                right--;
            }
            return true;
        }
    }
}

File: LeetCode/LeetCode101_150/125_Valid_Palindrome.cpp
#include <iostream>
#include <string>

#include "../..//LeetCodeCpp/LeetCode101_150.h"
#include "../..//LeetCodeCpp/Common.h"

using namespace std;

bool LeetCode::LeetCode101_150::isPalindrome(string s) {
    if (s.empty()) return true;
    int l = 0, r = s.size() - 1;
    while (l < r) {
        if (!isalnum(s[l])) ++l;
        else if (!isalnum(s[r])) --r;
        else if (tolower(s[l]) != tolower(s[r])) return false;
        else { ++l; --r; }
    }
    return true;
}